### PR TITLE
feat: implement Span32 struct

### PR DESCRIPTION
**Pull request by zarboq**
Span32 struct with a max size of 32

# Pull Request type

- Feature

Resolves: #305

### CODE REVIEW

File: src/reader/reader.cairo [hunk 1]
```diff
@@ -133,8 +134,8 @@ fn get_withdrawl(data_store: IDataStoreSafeDispatcher, key: felt252) -> Withdraw
         callback_contract: 0.try_into().unwrap(),
         ui_fee_receiver: 0.try_into().unwrap(),
         market: 0.try_into().unwrap(),
-        long_token_swap_path: ArrayTrait::new(),
-        short_token_swap_path: ArrayTrait::new(),
+        long_token_swap_path: Array32Trait::span(@ArrayTrait::new()),
```
Comment (enitrat) [inline src/reader/reader.cairo:137]: why not `Default::default().span32()`

File: src/utils/span32.cairo [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    ref serialized: Span<felt252>, mut curr_output: Array<T>, remaining: felt252
+) -> Option<Array<T>> {
+    if remaining == 0 {
+        return Option::Some(curr_output);
+    }
+    curr_output.append(TSerde::deserialize(ref serialized)?);
+    deserialize_array_helper(ref serialized, curr_output, remaining - 1)
+}
+
+impl SpanCopy<T> of Copy<Span32<T>>;
```
Comment (enitrat) [inline src/utils/span32.cairo:39]: derive Copy and Drop

File: src/utils/span32.cairo [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+// Local imports.
+use satoru::withdrawal::withdrawal::Withdrawal;
+
+// Span32.
+struct Span32<T> {
+    snapshot: Span<T>
+}
+
+fn serialize_array_helper<T, impl TSerde: Serde<T>, impl TDrop: Drop<T>>(
```
Comment (enitrat) [inline src/utils/span32.cairo:17]: derive Serde instead

File: src/utils/span32.cairo [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        return Option::Some(curr_output);
+    }
+    curr_output.append(TSerde::deserialize(ref serialized)?);
+    deserialize_array_helper(ref serialized, curr_output, remaining - 1)
+}
+
+impl SpanCopy<T> of Copy<Span32<T>>;
+impl SpanDrop<T> of Drop<Span32<T>>;
+
+impl Span32Serde<T, impl TSerde: Serde<T>, impl TDrop: Drop<T>> of Serde<Span32<T>> {
```
Comment (enitrat) [inline src/utils/span32.cairo:42]: derive Serde instead

File: src/utils/span32.cairo [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    }
+}
+
+trait Array32Trait<T> {
+    fn span(self: @Array<T>) -> Span32<T>;
+}
+
+impl Array32<T> of Array32Trait<T> {
+    #[inline(always)]
+    fn span(self: @Array<T>) -> Span32<T> {
```
Comment (enitrat) [inline src/utils/span32.cairo:100]:
```suggestion
    fn span32(self: @Array<T>) -> Span32<T> {
```

File: src/utils/span32.cairo [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    fn write(
+        address_domain: u32, base: StorageBaseAddress, value: Span32<ContractAddress>
+    ) -> SyscallResult<()> {
+        StoreContractAddressSpan::write_at_offset(address_domain, base, 0, value)
+    }
+
+    fn read_at_offset(
+        address_domain: u32, base: StorageBaseAddress, mut offset: u8
+    ) -> SyscallResult<Span32<ContractAddress>> {
+        let mut arr: Array<ContractAddress> = ArrayTrait::new();
```
Comment (enitrat) [inline src/utils/span32.cairo:122]: is it maybe easier to call `Store::<Span>::read_at_offset` and wrap the returned Span inside the Span32 struct?

File: src/utils/span32.cairo [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                },
+                Option::None(_) => {
+                    break Result::Ok(());
+                }
+            };
+        }
+    }
+
+    fn size() -> u8 {
+        255 * Store::<felt252>::size()
```
Comment (enitrat) [inline src/utils/span32.cairo:174]: the goal of Span32 is to have a size of 32 max :)

Comment (zarboq) [reply src/utils/span32.cairo:174]: oops

File: tests/data/test_withdrawal.cairo [hunk 8]
```diff
@@ -59,14 +60,12 @@ fn test_set_withdrawal_new_and_override() {
...
     // TODO make these Span32
-    let long_token_swap_path: Span<ContractAddress> = array![
-        1.try_into().unwrap(), 2.try_into().unwrap(), 3.try_into().unwrap()
-    ]
-        .span();
-    let short_token_swap_path: Span<ContractAddress> = array![
-        4.try_into().unwrap(), 5.try_into().unwrap(), 6.try_into().unwrap()
-    ]
-        .span();
+    let long_token_swap_path: Span32<ContractAddress> = Array32Trait::span32(
```
Comment (enitrat) [inline tests/data/test_withdrawal.cairo:63]: it doesnt work with `array![
        1.try_into().unwrap(), 2.try_into().unwrap(), 3.try_into().unwrap()].span32()`?

File: src/data/data_store.cairo [hunk 9]
```diff
@@ -662,19 +662,24 @@ mod DataStore {
 
             // Specifically handle case where there is only one withdrawal
             let last_withdrawal_index = withdrawals.len() - 1;
+            // @b how does this check that there is only 1 withdrawal
```
Comment (enitrat) [inline src/data/data_store.cairo:665]: remove this

File: src/data/data_store.cairo [hunk 10]
```diff
@@ -662,19 +662,24 @@ mod DataStore {
...
             if index == last_withdrawal_index {
                 withdrawals.pop_front();
                 self.withdrawal_indexes.write(key, 0);
                 self._remove_account_withdrawal(key, account);
                 return;
             }
 
-            let mut last_withdrawal_maybe = withdrawals.pop_front();
+            let mut last_withdrawal_maybe = withdrawals.pop_front(); // @b Take first withdrawal
```
Comment (enitrat) [inline src/data/data_store.cairo:673]: remove all comments

**Review by enitrat — COMMENTED** [review]
after that good

**Review by Sk8erboi84 — APPROVED** [review]

**Pull request merged**